### PR TITLE
[sailfish-secrets] Write documentation for plugin implementers. Contributes to JB#36797

### PR DESCRIPTION
**Pull request by chriadam**

### CODE REVIEW

**Comment by chriadam** [comment]
This is now ready for review.

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 1]
```diff
@@ -15,10 +15,642 @@
...
+  In order to implement a Crypto extension plugin, plugin implementers should
+  specify the following in their .pro file:
+  \code
+  CONFIG += link_pkgconfig
+  PKGCONFIG += sailfishcryptopluginapi
+  \endcode
+
+  The CryptoPlugin class extends the Sailfish::Secrets::PluginBase abstract
+  base class, and the sailfishcryptopluginapi pkgconfig file will entail
+  a dependency upon sailfishsecretspluginapi and ailfishsecrets, as
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: « ailfishsecrets », one 's' is missing

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 2]
```diff
@@ -15,10 +15,642 @@
...
+ */
+
+/*!
+ * \enum CryptoPlugin::EncryptionType
+ *
+ * This enum defines the types of encryption capability which may be offered by plugins
+ *
+ * \value NoEncryption No encryption is performed
+ * \value SoftwareEncryption Encryption is performed by "normal" rich execution environment application
+ * \value TrustedExecutionSoftwareEncryption Encryption is performed by trusted execution environment application
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: Maybe defining « trusted » here, I have no clue what is a normal rich and a trusted environment.

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 3]
```diff
@@ -15,10 +15,642 @@
...
+ *        into \a randomData for the caller identified by the given
+ *        \a callerIdent.
+ *
+ * If the plugin itself is locked, this function should return a
+ * Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Result::Failed and the error code set to
+ * Sailfish::Crypto::Result::CryptoPluginIsLockedError.
+ *
+ * The \a customParameters will contain plugin-specific parameters which may
+ * be required by the plugin.  Such parameters must be documented for clients
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: Two spaces.

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 4]
```diff
@@ -15,10 +15,642 @@
...
+ * passphrase fails to decrypt the key content, the plugin should return a
+ * Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Result::Failed and the error code set to
+ * Sailfish::Crypto::Result::CryptoPluginIncorrectPassphrase.
+ *
+ * The out-parameter \a keyMetadata should include only key metadata (such as
+ * its algorithm, size, and its full identifier) as well as public key data,
+ * but not private or secret key data.
+ *
+ * If the collection name specified in the identifier of the \a keyTemplate
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: There is no keyTemplate in importAndStoreKey() API…

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 5]
```diff
@@ -15,10 +15,642 @@
...
+ * does not specify a valid collection managed by the plugin, the plugin
+ * should return a Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Failed and the error code set to
+ * Sailfish::Crypto::Result::InvalidKeyIdentifier and the storage error code
+ * set to Sailfish::Secrets::Result::InvalidCollectionError.
+ *
+ * If the secret name specified in the identifier of the \a keyTemplate
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: Idem.

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 6]
```diff
@@ -15,10 +15,642 @@
...
+ * Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Result::Failed and the error code set to
+ * Sailfish::Crypto::Result::CryptoPluginIsLockedError.
+ *
+ * The \a customParameters will contain plugin-specific parameters which may
+ * be required by the plugin.  Such parameters must be documented for clients
+ * in the documentation provided with the plugin, and otherwise should be
+ * ignored by plugin implementers.
+ *
+ * If the collection name specified in the identifier of the \a keyTemplate
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: To much copy-paste, I guess, keyTemplate is not part of the API.

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 7]
```diff
@@ -15,10 +15,642 @@
...
+ * set to Sailfish::Secrets::Result::InvalidSecretError.
+ *
+ * If the plugin does not support retrieving key data, it should return a
+ * Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Failed and the error code set to
+ * Sailfish::Crypto::Result::UnsupportedOperation.
+ */
+
+/*!
+ * \fn CryptoPlugin::storedKeyIdentifiers(const QString &collectionName, QVector<Sailfish::Crypto::Key::Identifier> *identifiers)
```
Comment (dcaliste) [inline lib/CryptoPluginApi/extensionplugins.cpp]: There is no mention of what should be returned if collectionName is empty. I guess all.

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 8]
```diff
@@ -13,71 +13,1189 @@
...
+  PKGCONFIG += sailfishsecretspluginapi
+  \endcode
+ */
+
+/*!
+ * \fn PluginBase::name() const
+ * \brief Return the name of the plugin
+ *
+ * This name must be globally unique, so a fully-qualified-domain-name
+ * prefix is recommended.
```
Comment (dcaliste) [inline lib/SecretsPluginApi/extensionplugins.cpp]: Give an example here, it will be clearer.

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 9]
```diff
@@ -13,71 +13,1189 @@
...
 {
 }
 
+/*!
+ * \enum EncryptionPlugin::EncryptionType
+ *
+ * This enum defines the types of encryption capability which may be offered by plugins
+ *
+ * \value NoEncryption No encryption is performed
+ * \value SoftwareEncryption Encryption is performed by "normal" rich execution environment application
```
Comment (dcaliste) [inline lib/SecretsPluginApi/extensionplugins.cpp]: Same than in crypto, if possible explain the difference between normal rich and trusted.

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 10]
```diff
@@ -13,71 +13,1189 @@
...
+ * \value SecurePeripheralEncryption Encryption is performed by a secure hardware peripheral via TEE application
+ */
+
+/*!
+ * \enum EncryptionPlugin::EncryptionAlgorithm
+ *
+ * This enum defines the encryption algorithms which may be used by plugins
+ *
+ * \value NoAlgorithm No encryption is performed
+ * \value CustomAlgorithm Some custom encryption algorithm is used by the plugin
```
Comment (dcaliste) [inline lib/SecretsPluginApi/extensionplugins.cpp]: Thanks for this additional custom value ;)

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 11]
```diff
@@ -13,71 +13,1189 @@
...
+ *        out-parameter.
+ *
+ * If the plugin itself is locked, this function should return a
+ * Sailfish::Secrets::Result with the result code set to
+ * Sailfish::Secrets::Result::Failed and the error code set to
+ * Sailfish::Secrets::Result::SecretsPluginIsLockedError.
+ *
+ * If the given \a collectionName is empty, or is the special value
+ * "standalone", the plugin should return the names of standalone secrets
+ * which are stored in the storage managed by the storage plugin.
```
Comment (dcaliste) [inline lib/SecretsPluginApi/extensionplugins.cpp]: What should be the return value if collectionName is empty or standalone and the plugin does not support standalone secrets?

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 12]
```diff
@@ -13,71 +13,1189 @@
...
+ *
+ * If the plugin itself is locked, this function should return a
+ * Sailfish::Secrets::Result with the result code set to
+ * Sailfish::Secrets::Result::Failed and the error code set to
+ * Sailfish::Secrets::Result::SecretsPluginIsLockedError.
+ *
+ * If the given \a collectionName is empty, or is the special value
+ * "standalone", the plugin should return the names of standalone secrets
+ * which match the filter requirements which are stored in the storage
+ * managed by the storage plugin.
```
Comment (dcaliste) [inline lib/SecretsPluginApi/extensionplugins.cpp]: Idem than for getSecret().

**Comment by dcaliste** [comment]
Very nice documentation, especially about the expected return code on specific cases. I'll have to check my own plugin soon ;)

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 1 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: Thanks, fixed

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 2 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: TEE is an industry term for a secure area inside the main processor.  I've added the acronyms explicity now to make it more clear.

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 3 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: I tend to manually typeset with double space to emulate m-dash space equivalence after periods.  qdoc ignores such whitespace only, so it's only an issue for string literals or translation strings.

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 4 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: I will be fixing that API shortly

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 6 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: Thanks, fixed

File: lib/CryptoPluginApi/extensionplugins.cpp (hunk 7 above)
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: Empty should be treated as invalid.  Also the errorCode should be StorageError in this case, which I have fixed now also.  Thanks.

File: lib/SecretsPluginApi/extensionplugins.cpp (hunk 8 above)
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Done

File: lib/SecretsPluginApi/extensionplugins.cpp (hunk 9 above)
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Added the explicit acronyms.

File: lib/SecretsPluginApi/extensionplugins.cpp (hunk 10 above)
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: We should have a CustomWhatever in each enum.  If we don't that needs to be fixed - please let me know if you spot such a case.

File: lib/SecretsPluginApi/extensionplugins.cpp (hunk 11 above)
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Clarified that it should return success after setting the outparam to an empty list.

File: lib/SecretsPluginApi/extensionplugins.cpp (hunk 12 above)
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Done

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 13]
```diff
@@ -15,10 +15,642 @@
...
+ * Sailfish::Crypto::Result::Failed and the error code set to
+ * Sailfish::Crypto::Result::CryptoPluginIsLockedError.
+ *
+ * The \a customParameters will contain plugin-specific parameters which may
+ * be required by the plugin.  Such parameters must be documented for clients
+ * in the documentation provided with the plugin, and otherwise should be
+ * ignored by plugin implementers.
+ *
+ * The key template specifies some information about the desired key (such as
+ * its algorithm and size) as well as filter data which should be associated
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: You are correct, will clarify this

File: lib/CryptoPluginApi/extensionplugins.cpp [hunk 14]
```diff
@@ -15,10 +15,642 @@
...
+ *
+ * If the plugin itself is locked, this function should return a
+ * Sailfish::Crypto::Result with the result code set to
+ * Sailfish::Crypto::Result::Failed and the error code set to
+ * Sailfish::Crypto::Result::CryptoPluginIsLockedError.
+ *
+ * The \a customParameters will contain plugin-specific parameters which may
+ * be required by the plugin.  Such parameters must be documented for clients
+ * in the documentation provided with the plugin, and otherwise should be
+ * ignored by plugin implementers.
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/CryptoPluginApi/extensionplugins.cpp]: Yes, have clarified and added documentation for this case to all appropriate methods

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 15]
```diff
@@ -13,71 +13,1189 @@
...
+ *
+ * If the encrypted storage plugin supports storing collections of secrets,
+ * it must implement this method by returning the names of currently
+ * stored collections into the out-parameter \a names and returning
+ * a Sailfish::Secrets::Result with the result code set to
+ * Sailfish::Secrets::Result::Succeeded.
+ *
+ * Otherwise, it should write an empty list of collection names to the
+ * out-parameter \a names and return a Sailfish::Secrets::Result with
+ * the result code set to Sailfish::Secrets::Result::Succeeded.
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Yes, have added documentation for this to all appropriate methods

File: lib/SecretsPluginApi/extensionplugins.cpp [hunk 16]
```diff
@@ -13,71 +13,1189 @@
...
+ * If the plugin itself is locked, this function should return a
+ * Sailfish::Secrets::Result with the result code set to
+ * Sailfish::Secrets::Result::Failed and the error code set to
+ * Sailfish::Secrets::Result::SecretsPluginIsLockedError.
+ *
+ * If no collection with that name exists in the storage managed by the plugin,
+ * the plugin should return a Sailfish::Secrets::Result
+ * with the result code set to Sailfish::Secrets::Result::Failed and the
+ * error code set to Sailfish::Secrets::Result::InvalidCollectionError.
+ *
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/SecretsPluginApi/extensionplugins.cpp]: Clarified that it should return InvalidSecretError

File: lib/Secrets/doc/sailfish-secrets-plugins.qdoc [hunk 17]
```diff
@@ -92,7 +92,9 @@ protected secret in the default storage plugin:
 \code
 // Define a standalone secret (no collection name specified in the identifier)
 Sailfish::Secrets::Secret standaloneSecret(
-        Sailfish::Secrets::Secret::Identifier("StandaloneSecret"));
+        Sailfish::Secrets::Secret::Identifier("StandaloneSecret"),
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Secrets/doc/sailfish-secrets-plugins.qdoc]: fixed, thanks

File: lib/Secrets/doc/sailfish-secrets-plugins.qdoc [hunk 18]
```diff
@@ -92,7 +92,9 @@ protected secret in the default storage plugin:
...
+        QString(),
+        Sailfish::Secrets::SecretManager::DefaultStoragePluginName);
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Secrets/doc/sailfish-secrets-plugins.qdoc]: Indeed, thanks

**Comment by chriadam** [comment]
Thanks very much for the comprehensive review!  I've now fixed according to review comments, and rebased.  Merging now.

**Pull request merged**